### PR TITLE
Adding scrollTop and scrollLeft helpers

### PR DESCRIPTION
**Pull request by alexander-alvarez**
Taken from https://github.com/html-next/vertical-collection/blob/master/tests/helpers/wait.js and used in https://github.com/alphasights/ember-scrollable/pull/74

Related to https://github.com/alphasights/ember-scrollable/pull/57 and don't need https://github.com/rwjblue/ember-lifeline/pull/19

**tl;dr**
From my understanding `requestAnimationFrame` causes a paint cycle to be triggered, which flushes out the scroll events that have been fired (which otherwise would have been evaluated asyncronously up to the browser)

### CODE REVIEW

**Comment by cibernox** [comment]
I think that it would make more sense to have a `scrollTo(selectorOrHTMLElement, value)` helper that does both things, do the scroll and return the promise.

You could then use it in your tests like this:

```js
import { scrollTo } from 'ember-native-dom-helpers';

test('Load more data with infinite scroll', async function(assert) {

  this.render(...);
  assert.equal(findAll('.item').length, 10);
  await scroll('.item-list', 100);
  assert.equal(findAll('.item').length, 20, 'Ten more items have been loaded'); 
});
```

Makes sense? I think it reads nicer and is aligned with the idea that other interaction helpers return a promise that fulfils when all async stuff has finished.

**Comment by cibernox** [comment]
This is a great adition BTW. I never had to test asyncronous behavious triggered by scroll but with this it would be a no brainer.

**Comment by alexander-alvarez** [comment]
Not a problem

**Comment by alexander-alvarez** [comment]
Made two helpers, `scrollTop` and `scrollLeft` to mimic the properties used to trigger scrolls on an `HTMLElement`

**Comment by alexander-alvarez** [comment]
test failures are seemingly unrelated

File: addon-test-support/scroll.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+/*
+  Triggers a paint (and therefore flushes any queued up scroll events).
+
+  @method triggerFlushWithPromise
+  @return {RSVP.Promise}
+  @private
+*/
+function triggerFlushWithPromise(){
+  return wait().then(() => {
```
Comment (cibernox) [inline addon-test-support/scroll.js]: I'm not familiar with the problem, but intuitively I feel that the order should be the opposite. My understanding is that the scroll event is not fired until the next time the screen is painted. As a consequence of that scroll event, some javascript will be executed and maybe some asynchronous behaviour will be triggered, and we should wait for it.

It seems to me that first we should wait for the `requestAnimationFrame` so the event is fired and then `wait` for whatever work that event handler started.

File: addon-test-support/scroll.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+/*
+  Scrolls DOM element or selector to the given position (if the DOM element is currently overflowed).
+  The promise resolves after the scroll event has been triggered.
+
+  @method scrollTop
+  @param {String|HTMLElement} selector
+  @param {Number} position
+  @return {RSVP.Promise}
+  @public
+*/
```
Comment (cibernox) [inline addon-test-support/scroll.js]: I'm thinking that a single `scrollTo` (like `window.scrollTo`) method that takes X and Y coordinates would be simpler. I think that it's not so uncommon to want to scroll both axis and wait, and with separate methods users would have to do

```
let scrollYPromise = scrollTop('.container', 500);
let scrollXPromise = scrollLeft('.container', 200);
await Ember.RSVP.all(scrollYPromise, scrollXPromise);
```

I don't feel very strong about this tho, but I think that `await scrollTo(200, 500);` is conceptually simpler that having two methods, even if most of the time scroll is only vertical.

File: addon-test-support/scroll.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+/*
+  Triggers a paint (and therefore flushes any queued up scroll events).
+
+  @method triggerFlushWithPromise
+  @return {RSVP.Promise}
+  @private
+*/
+function triggerFlushWithPromise(){
```
Comment (cibernox) [inline addon-test-support/scroll.js]: `paintAndWait` perhaps?

File: addon-test-support/scroll.js (hunk 2 above)
Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: fair enough... I can buy this :+1: 
updated API will be like so
`scrollTo(element, verticalPosition, horizontalPosition)`

File: addon-test-support/scroll.js (hunk 1 above)
Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: In all honesty, I understand it enough to rationalize it in my head, not defend it -- so that's my bad.
@pzuraq can you chime in here? That way we add the explanation to the documentation and we all can understand why

File: addon-test-support/scroll.js (hunk 2 above)
Comment (cibernox) [inline addon-test-support/scroll.js]: I think that it should be the opposite to match `window.scrollTo(horizontalScroll, verticalScroll)`.

Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: oh that's annoying... horizontal scroll is significantly less common than vertical scroll, so I was hoping if we had the horizontal argument last we could get away with omitting it most of the time.
```js
let vOffset = 10;
scrollTo(elem, vOffset);
// vs
scrollTo(elem, 0, vOffset);
```
this makes me lean towards using separate methods like jQuery does https://api.jquery.com/scrollTop/ & https://api.jquery.com/scrollLeft/

Comment (cibernox) [inline addon-test-support/scroll.js]: Yeah, I know that is slightly annoying, but I still think that we should copy DOM apis. To some extent I think that all functions that take x and y values, should take them in that order. It's the way we speak. 

If people complain we can create a `scrollTo` helper that would be just` scrollTop(el, y) { return scrollTo(el, 0, y) }`.

Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: I think we should be copying the setter methods `Element.scrollTop` and `Element.scrollLeft` instead of `Window.scrollTo`

File: addon-test-support/scroll.js (hunk 1 above)
Comment (pzuraq) [inline addon-test-support/scroll.js]: So the reason this helper exists is because we actually wrap our usage in of requestAnimationFrame `vertical-collection` with Embe.run: https://github.com/html-next/vertical-collection/blob/master/addon/-private/scheduler/index.js. 

So in order to fully clear out the runloop, we actually need to wait for the next RAF to finish. I'm not sure if this will be helpful to most Ember users, since most of them will just be scheduling things on the standard runloop and there isn't a standard Ember way to do RAF (thus the hack-y scheduler we built).

File: addon-test-support/scroll.js (hunk 2 above)
Comment (cibernox) [inline addon-test-support/scroll.js]: And we can, but I think that they should be built on top of a method of a this.

File: addon-test-support/scroll.js (hunk 1 above)
Comment (cibernox) [inline addon-test-support/scroll.js]: @pzuraq I'd lie if I told your that I got it. So in short, is this kind of approach needed for people adding `scroll` events that trigger async behaviour?
If so, which one is more correct?
```js
return new RSVP.Promise(function(resolve) { window.requestAnimationFrame(resolve) }).then(() => wait())
```
or
```js
return wait().then(() => new RSVP.Promise(function(resolve) { window.requestAnimationFrame(resolve) }));
```

Comment (pzuraq) [inline addon-test-support/scroll.js]: So I think `requestAnimationFrame` shouldn't be necessary at all here, and a standard `wait` should be enough. I'm not 100% sure on that, but the goal for us was to make sure that we captured all possible runloops, and we basically added an extra runloop that happens inside of a RAF, thus the need to schedule an extra RAF to clear the queue.

I'll see if I can play around and try to make sure that's true, there are some weird edge cases with scroll events in general.

Comment (cibernox) [inline addon-test-support/scroll.js]: I thought that the requestAnimationFrame because scroll events are asynchronous and don't fire immediately after setting the scroll 🤔

Comment (pzuraq) [inline addon-test-support/scroll.js]: yeah, that's what I want to play around with. If a user sends a scroll event and we `wait` will that be long enough to trigger it? Not entirely sure.

For our tests, the RAF is actually triggered _by_ the scroll event, so it couldn't possibly be what ends up triggering it. There may be a better way to make sure the event gets triggered though, maybe manually.

Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: I can go back and test, but in ember-scrollable I couldn't get the scroll events to be sychronous for the life of me, even using the wait helper -- the wait helper is really simply waiting an arbitrary amount of time, the browser decides when to trigger the scroll.

Then I plugged this in and it made unicorns appear. I can investigate again to make sure I wasn't going crazy and report back.

Comment (pzuraq) [inline addon-test-support/scroll.js]: Yeah I believe it, but like I said I don't think this is specified behavior necessarily. Basically we're relying on the browser's queue's the line up just right, and I'm not sure that will always be the case (you can see [an example of this here](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/), although they don't go into UI Events and I can't seem to find where those are scheduled in the spec).

I think a better strategy may be to use [dispatchEvent](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent) to synchronously send the scroll event just after scrollTop/scrollLeft is modified.

Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: from slack:
> see https://github.com/alphasights/ember-scrollable/compare/master...alexander-alvarez:testing-wait?expand=1
the first commit with just wait fails tests

From my investigation it looks like we don't even need the wait helper

Comment (pzuraq) [inline addon-test-support/scroll.js]: Yes, it definitely works. I just don't know if that's _guaranteed_ in all browsers, everywhere, forever. The reason `wait` works at all is because Ember Promises are actually scheduled on the runloop, as are UI Events. RAF steps _outside_ of the normal runloop system, so we're relying on the browser's behavior and I can't figure out where UI Events are scheduled relative to RAF.

We may want to loop @runspired in on this because he's more familiar with how the browser's internal scheduling works.

Comment (cibernox) [inline addon-test-support/scroll.js]: @pzuraq actually that's not true. `wait` doesn't work for promises to resolve. It waits for 
1. jQuery's ajax request (not fetch, not raw XHR)
2. The runloop to be totally empty
3. Waiters. Waiters are functions developers or addons can register that can do absolutely anything. A waiter function is pending if it returns a falsy value.

`wait` might not not needed to wait for the event to fire, but we must `return wait()` to allow users to wait for any **asynchronous consequence** of that event.

Comment (pzuraq) [inline addon-test-support/scroll.js]: Ah, my bad, I thought it was just the Promise resolution it waited for. Been a while since I looked at that code 😅 

Either way though, the point I'm trying to make is that this may not be guaranteed. It works for now, definitely, and it looks like we're actually relying on that behavior in the tests for `vertical-collection`. The other option would be to send the event synchronously using `dispatchEvent`, like I mentioned before, and I'm not sure which one would be better here.

Comment (runspired) [inline addon-test-support/scroll.js]: setting scrollTop is synchronous. The scroll event caused by setting it is fired off almost immediately, but you would need to wait for it.  It would be captures by any `setTimeout` based waiter though, which already exists iirc.

Comment (cibernox) [inline addon-test-support/scroll.js]: in that case, seems that we don't need rAF. Which is good news for IE9

Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: chiming in here. I haven't seen scroll be synchronous, just pulled this from putting a breakpoint [here](https://github.com/alphasights/ember-scrollable/compare/master...alexander-alvarez:testing-wait#diff-bcc853db8ecac305500add8bb7a0f313R164) and running tests in Chrome
image

Comment (cibernox) [inline addon-test-support/scroll.js]: @alexander-alvarez Indeed `scroll (async)` seems to suggest otherwise.

Comment (runspired) [inline addon-test-support/scroll.js]: I didn't say scroll was sync, I said setting scrollTop was sync, which generates an almost immediate scroll event. The following pattern should always work:

```
function updateScrollAndWaitForEvent(element, value) {
   return new Promise(resolve => {
    element.scrollTop = value;
    setTimeout(resolve, 1);
  });
}
```

Comment (alexander-alvarez) [inline addon-test-support/scroll.js]: this falls in line with what I was observing in https://github.com/alphasights/ember-scrollable/compare/master...alexander-alvarez:testing-wait?expand=1

do we care about the subtle difference between `setTimeout` and `requestAnimationFrame`?
i.e. https://developers.google.com/web/fundamentals/performance/rendering/optimize-javascript-execution#use_requestanimationframe_for_visual_changes
